### PR TITLE
Dynamic work-in-progress in Parallel Console

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
See https://github.com/gradle/gradle/issues/1833.

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [x] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Branches_Checkpoints&branch_Gradle_Branches_Checkpoints=dl-issue-1833)
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 1]
```diff
@@ -103,87 +99,59 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
             unusedProgressLabels.push(progressArea.getBuildProgressLabels().get(i));
         }
-
-        // Try to empty the unassigned progress operations
-        while (!unusedProgressLabels.isEmpty() && !unassignedProgressOperations.isEmpty()) {
-            attach(unassignedProgressOperations.pop());
-        }
     }
 
     private void attach(ProgressOperation operation) {
```
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: I restructured the code in this file to make it easier to maintain but the diff is just horrible. Maybe using split diff will be easier for the review.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 2]
```diff
@@ -208,4 +176,48 @@ void renderNow() {
...
+                        parentIdToChildrenIds.put(parentId, children);
+                    }
+                    children.add(childId);
+                }
+
+                @Override
+                public void remove(OperationIdentifier childId) {
+                    Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
+                    if (children == null) {
+                        throw new IllegalStateException("");
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: :x: We should figure out how this'll work given my changes. We may just want to `return;` here.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 3]
```diff
@@ -103,87 +99,59 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
-        // Try to empty the unassigned progress operations
-        while (!unusedProgressLabels.isEmpty() && !unassignedProgressOperations.isEmpty()) {
-            attach(unassignedProgressOperations.pop());
-        }
     }
 
     private void attach(ProgressOperation operation) {
         // Skip attach if a children is already present
-        if (isChildAssociationAlreadyExists(operation.getOperationId())) {
+        if (registry.find(operation.getOperationId()).hasChildren()) {
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: :o: For later: consider how we'll handle things like parallel test execution where we want to display different branches of the tree

File: subprojects/logging/src/test/groovy/org/gradle/internal/logging/console/ConsoleFunctionalTest.groovy [hunk 4]
```diff
@@ -55,7 +55,7 @@ class ConsoleFunctionalTest extends Specification {
         then:
         ConcurrentTestUtil.poll(1) {
             assert statusBar.display == '<---> 0% INITIALIZING [0s]'
-            assert progressArea.display == [IDLE, IDLE, IDLE, IDLE]
+            assert progressArea.display == [IDLE]
```
Comment (eriwen) [inline subprojects/logging/src/test/groovy/org/gradle/internal/logging/console/ConsoleFunctionalTest.groovy]: :x: Let's add a test that ensures that after 2 workers spawn, they go back to `IDLE` and do not, in fact, go away or blank.

File: subprojects/logging/src/test/groovy/org/gradle/internal/logging/console/ConsoleFunctionalTest.groovy [hunk 5]
```diff
@@ -199,19 +199,48 @@ class ConsoleFunctionalTest extends Specification {
...
         renderer.onOutput(startEvent(3, ':bar'))
         renderer.onOutput(startEvent(4, ':baz'))
-        renderer.onOutput(startEvent(5, ':nope'))
+        renderer.onOutput(startEvent(5, ':foz'))
+        renderer.onOutput(startEvent(6, ':nope'))
 
         then:
         ConcurrentTestUtil.poll(1) {
-            assert progressArea.display == ['> :wat', '> :foo', '> :bar', '> :baz']
+            assert progressArea.display == ['> :wat', '> :foo', '> :bar', '> :baz', '> :foz']
```
Comment (eriwen) [inline subprojects/logging/src/test/groovy/org/gradle/internal/logging/console/ConsoleFunctionalTest.groovy]: :o: maybe assert that the console height is 10 here to make the test clear

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 3 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: I think it won't be a problem, any operation with no children will be displayed. A child operation _pause_ the execution of a parent operation. If this assumption is wrong, then we will need to change the logic here.

**Comment by lacasseio** [comment]
Let's not merge this right away. With the rebase to latest master, a lot of work-in-progress labels are shown but are all kept `> IDLE`. I make me think we are trying to display progress event without any displayable strings.

**Comment by eriwen** [comment]
@lacasseio I'm pretty sure I know why. This is because now a bunch of progress loggers are being created without "progressDescriptions", that is, short descriptions.

**Comment by eriwen** [comment]
@lacasseio This bug is not a result of your change. @bmuschko's and my changes from yesterday conflict here, causing it. We 3 should solve it together, but know that it's not a result of your PR here (probably).

**Comment by lacasseio** [comment]
@eriwen You are right. I have a solution that will work in 2 ways. 1) for each batch of events, if the complete event is in the current batch, ignore the event. 2) If the start event can't be displayed (doesn't have a short description), ignore the event.

**Comment by eriwen** [comment]
@lacasseio A couple issues I noticed trying this out:
 - Sometimes we don't have anything to display on the console for some reason
 - it doesn't seem to respect console height fully

screenflow

**Comment by lacasseio** [comment]
@eriwen I will look at the first point. As for the second point, this means we should count the build status and cursor parking rows? Basically, should the algorithm for `ConsoleLayoutCalculator` be `(consoleMetaData.getRows() / 2) - 2` (for your example, this would be 1 line in work-in-progress) or be `(consoleMetaData.getRows() - 2) / 2` (for your example, this would be 2 lines in work-in-progress)?

**Comment by eriwen** [comment]
@lacasseio I filed #1980 for the 2nd one

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 6]
```diff
@@ -103,86 +129,79 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
         }
     }
 
-    private boolean isChildAssociationAlreadyExists(OperationIdentifier parentId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children != null && !children.isEmpty()) {
-            return true;
+    // Any ProgressOperation in the parent chain has a message, the operation is considered renderable.
+    private boolean isRenderable(ProgressOperation operation) {
+        for (ProgressOperation current = operation;
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: :x: We should reuse/unify the logic here with how we handle this in `BuildStatusRenderer`

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 7]
```diff
@@ -208,4 +227,48 @@ void renderNow() {
             label.setText(labelFormatter.format(operation));
         }
     }
+
+    private static class ParentRegistry {
+        private final Map<OperationIdentifier, Set<OperationIdentifier>> parentIdToChildrenIds = new HashMap<OperationIdentifier, Set<OperationIdentifier>>();
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: ⭕️  Is there an opportunity to add functionality to `ProgressOperations` instead of maintaining the progress operation hierarchy 2 ways?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 8]
```diff
@@ -208,4 +227,48 @@ void renderNow() {
             label.setText(labelFormatter.format(operation));
         }
     }
+
+    private static class ParentRegistry {
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: :o: We might break this out and test it separately

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 9]
```diff
@@ -75,23 +77,47 @@ public void onOutput(OutputEvent event) {
...
     }
 
     @Override
     public void onOutput(Iterable<OutputEvent> events) {
-        super.onOutput(events);
+        Set<OperationIdentifier> completeEventOperationIds = toOperationIdSet(toProgressCompleteEvents(events));
+        Set<OperationIdentifier> startEventOperationIdsToSkip = new HashSet<OperationIdentifier>();
+
+        for (OutputEvent event : events) {
+            if (event instanceof ProgressStartEvent && completeEventOperationIds.contains(((ProgressStartEvent) event).getProgressOperationId())) {
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: This achieves one of our other stories for polishing console output. I feel like this is orthogonal to the changes we made recently, though, so I would be surprised if this fixed all the issues we are seeing. 

Let's reconsider if this is really better as a separate change.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 6 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: I'm unsure how this can be unified with `BuildStatusRenderer`. Could you help me see through this?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 8 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: This was merged with `ProgressOperation` and `ProgressOperations`.

**Comment by eriwen** [comment]
@lacasseio we should probably rebase this on latest `release` and change the PR branch base, too

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 10]
```diff
@@ -131,96 +124,58 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
+        // No more unused label? Try to resize.
+        if (unusedProgressLabels.isEmpty()) {
+            int newValue = operationIdToAssignedLabels.size() + 1;
+            resizeTo(newValue);
         }
 
-        if (association == null) {
-            unassignedProgressOperations.addLast(operation);
+        // Try to use a new label
+        if (unusedProgressLabels.isEmpty()) {
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: :o: We're doing the same check as above. Can we just inline `resizeTo(oITAL.size() + 1)` here?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 11]
```diff
@@ -131,96 +124,58 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
-    private void removeDirectChildOperationId(OperationIdentifier parentId, OperationIdentifier childId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children == null) {
-            return;
-        }
-        children.remove(childId);
-        if (children.isEmpty()) {
-            parentIdToChildrenIds.remove(parentId);
+        if (operation.getParent() != null) {
+            attach(operation.getParent());
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: Do we need an `isRenderable()` check here?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java [hunk 12]
```diff
@@ -131,96 +124,58 @@ private void resizeTo(int newBuildProgressLabelCount) {
...
+        } else if (!unassignedProgressOperations.isEmpty()) {
+            attach(unassignedProgressOperations.pop());
```
Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: Do we need an `isRenderable()` check here too?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 10 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: We may have or have not resized the WIP area. If we didn't resize, we need to put the progress operation in the `unassignedProgressOperations` list.

Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: got it. Can we add a comment that this duplicate check is intentional?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 11 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: Great catch. I have created a test and fixed the code.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java (hunk 12 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: Here we shouldn't have any non-renderable item in that list as we ignore any non-renderable operation in `attach`

Comment (eriwen) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java]: 👍
